### ISSUE TITLE
Conjugate gradients method do not converge for large values of the regularization [CPU/GPU]

### ISSUE THREAD

**Issue by lucidyan** [benfred/implicit#106]
```python
import os
os.environ["OPENBLAS_NUM_THREADS"] = "1"

import numpy as np
import implicit
import scipy.sparse

Cui = scipy.sparse.random(
    m=10000, 
    n=100, 
    density=0.0005,
    format='coo', 
    dtype=np.float32, 
    random_state=42, 
    data_rvs=None
)

model = implicit.als.AlternatingLeastSquares(
    factors=64,
    regularization=10,
    dtype=np.float32,
    use_native=True,
    use_cg=True,
    use_gpu=False,
    iterations=10,
    calculate_training_loss=True,
    num_threads=0
)
model.fit(Cui.T)
model.training_loss_history
```
Output:

> [0.00025055279047685153,
 0.0002505180536281105,
 0.0002505180290097362,
 0.00025051802898084217,
 0.00025051802898079907,
 nan,
 nan,
 nan,
 nan,
 nan]

If we set the regularization to zero or change to the usual method of ALS, then the method converges and the loss does not become NaN

Same behaviour on GPU.

[SYSTEM INFORMATION]
Python 3.6.5
<details>
 <summary>python packages list</summary>
asn1crypto==0.24.0<br />backcall==0.1.0<br />bleach==2.1.3<br />catfish==1.4.4<br />certifi==2018.1.18<br />chardet==3.0.4<br />command-not-found==0.3<br />cryptography==2.1.4<br />cupshelpers==1.0<br />decorator==4.3.0<br />defer==1.0.6<br />distro-info==0.18<br />entrypoints==0.2.3<br />Guake==3.0.5<br />h5py==2.7.1<br />html5lib==1.0.1<br />httplib2==0.9.2<br />idna==2.6<br />implicit==0.3.5<br />ipykernel==4.8.2<br />ipython==6.4.0<br />ipython-genutils==0.2.0<br />ipywidgets==7.2.1<br />jedi==0.12.0<br />Jinja2==2.10<br />jsonschema==2.6.0<br />jupyter==1.0.0<br />jupyter-client==5.2.3<br />jupyter-console==5.2.0<br />jupyter-core==4.4.0<br />keyring==10.6.0<br />keyrings.alt==3.0<br />language-selector==0.1<br />launchpadlib==1.10.6<br />lazr.restfulclient==0.13.5<br />lazr.uri==1.0.3<br />lightdm-gtk-greeter-settings==1.2.2<br />MarkupSafe==1.0<br />menulibre==2.2.0<br />mistune==0.8.3<br />mugshot==0.4.0<br />nbconvert==5.3.1<br />nbformat==4.4.0<br />notebook==5.5.0<br />numpy==1.14.3<br />oauth==1.0.1<br />olefile==0.45.1<br />onboard==1.4.1<br />pandas==0.23.0<br />pandocfilters==1.4.2<br />parso==0.2.1<br />patsy==0.5.0<br />pbr==3.1.1<br />pexpect==4.5.0<br />pickleshare==0.7.4<br />Pillow==5.1.0<br />prompt-toolkit==1.0.15<br />psutil==5.4.2<br />ptyprocess==0.5.2<br />pycairo==1.16.2<br />pycrypto==2.6.1<br />pycups==1.9.73<br />pycurl==7.43.0.1<br />Pygments==2.2.0<br />pygobject==3.26.1<br />python-apt==1.6.0<br />python-dateutil==2.7.3<br />python-debian==0.1.32<br />python-debianbts==2.7.2<br />python-magic==0.4.16<br />pytz==2018.4<br />pyxdg==0.25<br />PyYAML==3.12<br />pyzmq==17.0.0<br />qtconsole==4.3.1<br />reportlab==3.4.0<br />requests==2.18.4<br />requests-unixsocket==0.1.5<br />scipy==1.1.0<br />screen-resolution-extra==0.0.0<br />SecretStorage==2.3.1<br />Send2Trash==1.5.0<br />sgt-launcher==0.2.4<br />simplegeneric==0.8.1<br />simplejson==3.13.2<br />six==1.11.0<br />ssh-import-id==5.7<br />statsmodels==0.9.0<br />systemd-python==234<br />terminado==0.8.1<br />testpath==0.3.1<br />tornado==5.0.2<br />tqdm==4.23.3<br />traitlets==4.3.2<br />ubuntu-drivers-common==0.0.0<br />ufw==0.35<br />unattended-upgrades==0.1<br />urllib3==1.22<br />virtualenv==15.1.0<br />wadllib==1.3.2<br />wcwidth==0.1.7<br />webencodings==0.5.1<br />widgetsnbextension==3.2.1<br />xcffib==0.5.1<br />xkit==0.0.0<br />zope.interface==4.3.2<br />
</details>

[SYSTEM]
CPU: Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz
OS: Ubuntu 18.04 LTS

[GPU]
Device: GeForce GTX 1080 Ti
Nvidia Driver Version: 390.48 
Cuda compilation tools: release 9.0, V9.0.176

**Comment by benfred** [comment]
Thanks for the bug report! I really appreciate the simple script to reproduce the problem.

I fixed this in #107. The issue seems to be with regularization like that the best solution contains users with a 0 valued vector for their factors. I've added an extra to check to prevent NaN values from getting inserted here - with this change the results for the cholesky optimizer and CG optimizer should be the same again.

This will be in the next release (v0.3.6) - which should be out later this week.
